### ISSUE TITLE
fod2fixel -peaks overestimates certain peak amplitudes

### ISSUE THREAD

**Issue by thijsdhollander** [MRtrix3/mrtrix3#1204]
First discovered on `master`, but also confirmed on `dev`.  All the below is using `dev`, since a few added features to `fod2fixel` in `dev` make this easier to illustrate.  **I marked this as "critical" since it directly impacts the behaviour of `dwi2response tournier`.**  I actually discovered this while playing around with some things related to `dwi2response tournier`.  I'll try to keep the description of the bug as short as possible, but I invite everyone to confirm that I'm not getting things wrong here.

Ok, so, as stated, I simplified things here; but this kind of problem happens in any possible setting (with(out) different options compared to this example) when using the `-peaks` option to `fod2fixel`.

So, simplified goal of this example here: just get the amplitude of the largest peak of all FODs.  First up, doing this with `sh2peaks -num 1` and `peaks2amp` works without issues.  These are the FODs I worked on, combined with those largest peaks using the fixel visualisation tool:

![screen0101](https://user-images.githubusercontent.com/11952694/33417378-5999c92a-d5f5-11e7-825f-13b8d2bce145.png)

Scaling is set the same for both, which causes the fixels/peaks to appear just slightly larger than the FOD peaks (not a problem with the actual data of either FODs or peaks, this is just what the viewer does with those scales in practice).  The hot colour scale for the peaks shows (in addition to their visual length) also encodes the peak amplitude.  I'll get to the grey scale voxels further on.

Next, trying to do exactly the same with `fod2fixel`.  I used `fod2fixel` with the `-peak` option output, as well as the `-fmls_no_thresholds`, `-maxnum 1` and `-dirpeak` options; but the essential problem/bug happens just as well without those latter 3 options.  Showing the fixels/peaks again combined with the same FODs as before (length as well as hot colour encoding the `-peak` output fixel datafile):

![screen0102](https://user-images.githubusercontent.com/11952694/33417563-22995962-d5f6-11e7-85f7-2ab947f7ec0f.png)

Same scale as before was used; and you can clearly see the problem: most peaks are significantly larger... The grey scale voxels encode the ratio of these peak amplitudes to the former ones obtained by `sh2peaks`.  Most peaks amplitudes are overestimated by about a factor 1.35-ish.  Note the close-to-black voxels: here, the peaks aren't overestimated so much, and the ratio gets close(r) to 1.  At first sight, these seem to be the peaks that are aligning (more) closely to the z-axis (image shown is a coronal view).

Next, here's a scatter plot all (largest) peak amplitudes in this dataset; horizontal axis is the amplitude from `sh2peaks` (followed by `peaks2amp`), vertical axis is the amplitude from `fod2fixel -peaks` (followed by `fixel2voxel`):

![screen105](https://user-images.githubusercontent.com/11952694/33417840-2af2270a-d5f7-11e7-9265-4bc9ded6051e.png)

Dashed red line is the identity line.  Colour scale is the ratio of the amplitudes (same as the grey scale of voxels before).  Note the overestimation maxes out around 1.35-something, and is unrelated to the actual amplitude of the peak.  Also, most amplitudes are overestimated in this 1.35-something range, and only few are close to the identity line.

Final plot shows the overestimation ratio (y-axis) in function of the absolute value of the z-coordinate of the (unit length) peak orientation:

![screen106](https://user-images.githubusercontent.com/11952694/33418119-78942016-d5f8-11e7-8bf9-238f0837e962.png)

This confirms that peak orientations close to the z-axis cause less, to ultimately no, overestimation of the peak amplitude.  It doesn't seem to explain everything though (but there could also be some noise and/or numerical issues on the ratio too).

Of all commands used above, the problem seems to ultimately lie with `fod2fixel -peaks`.  As mentioned, this currently has direct implications for the accuracy/performance of `dwi2response tournier`.

If/when the cause is found and fixed, I reckon this calls for a hotfix with a tag straight to `master`... `dev` itself is definitely not ready yet to go to `master`, with outstanding discussions on renaming of certain commands/option names to be finished up first.  But it wouldn't be great to let users on `master` continue with what might be a wrong behaviour of `dwi2response tournier`.  Well, happy to hear everyone's opinion on this of course.  I'm currently not in a position to spend more time on the fix...  I hope the above findings may already provide insights in where to look for fixes...

**Comment by jdtournier** [comment]
OK, problem confirmed. I generated a little FOD field (code below if you're interested), then:
```
fod2fixel deltas.mif -peak peaks.mif fixels
fixel2voxel fixels/peaks.mif split_data fod2fixel.mif
```
and below is the resulting image of the amplitude of the first peak, with the ODFs overlaid:

![screenshot from 2017-11-30 16-24-26](https://user-images.githubusercontent.com/3221000/33441749-51f828f6-d5eb-11e7-818c-86f6f0b9f7f0.png)

So there does seem to be a problem...

I also note that `sh2peaks` fails when the peak is exactly aligned with the z-axis - I'll investigate.

**Comment by thijsdhollander** [comment]
> OK, first off, I don't expect the results of sh2peaks and fod2fixel -peaks to match - they perform fundamentally different operations.

I think you're confusing a few things.  I'll clarify, so we're certainly on the same front here.

> sh2peaks does exactly what it says,

Correct, no confusion there.

> but fod2fixel segments the fODF into fixels,

Still correct, but...

> and the -peaks option produces the mean orientation for that fixel, scaled by its integral over the fixel - there is no guarantee that the direction matches the actual peak exactly (depends on whether there are any asymmetries in the peak), or that the amplitude matches that of the peak (depends on how broad the peak is - narrower will tend to have larger amplitude for similar integral).

No no no, you're getting a bit off track here. :wink:  There's 2 things for each segmented fixel that get output, and they're in many ways independent from each other (so shouldn't be mixed up):

* The direction for the segmented fixel, stored in `directions.mif`: by default an average orientation for the fixel that indeed doesn't necessarily align with the "peak" (local maximum amplitude), but using the `-dirpeak` option, you can get the direction of the "peak" (local maximum amplitude) *instead*.  I used that option in my example above to make things (visually, for the fixel plot) as similar as possible to the `sh2peaks` output, but it doesn't even really matter at all here because...
* ...completely independent from the `direction.mif` output for each fixel, there's a series of fixel datafiles that can get output.  They've got nothing to do with the kind of direction output chosen, but output some metric for each fixel:
  * `-afd`: the integral you've mentioned above, but that's the `-afd` option, not the `-peak` option we're talking about here :wink: 
  * `-peak`: output the "peak" (local maximum) amplitude for the fixel.  So that's not the amplitude of the direction that gets out put per se.  It will be (well, _should_ be) when the `-dirpeak` option is used.  But regardless of specifying `-dirpeak` or not, `-peak` outputs the *peak* amplitude, which is/should be the same as what `sh2peaks` delivers; up to maybe small numerical differences of course.
  * `-disp`: a metric of dispersion, ratio of the 2 previous ones

So for most "large" and well defined fixels, the peak amplitude output via `-peak` should match the `sh2peaks` amplitude.  Not that it's relevant here (so just a side note), but note that if `fod2fixel -peak` would've output the amplitude along any other direction than that of the genuine peak, it would've appeared to underestimate the peak amplitude.  However, the problem we're observing here is one of *overestimation*.

So well, about the problem at hand then: I don't have the time to really dig into the code, but the suspicious relation of this bug/bias to the z-axis alignment potentially suggests something to do with the evaluation of the SH-basis to obtain this particular amplitude, or even more specifically with the dependence on the inclination angle (theta).  Maybe a problem somewhere with conversion from/to spherical polar coordinates...?

I haven't checked the `-afd` (integral) output yet with respect to this, as my focus was specifically on obtaining the peak amplitude (only)... but at this stage I'm praying the `-afd` output isn't affected by this, because that would have a much more problematic impact on a lot of things... :cold_sweat:

**Comment by thijsdhollander** [comment]
Ok, quickly checked the `-afd` output just to be sure; I don't think I'd sleep well if I wasn't sure that one works correctly... :fearful:  The good news is that one's ok.  In line with all the above, it's specifically the `-peak` output that is affected; and the `-disp` output consequently as well (in line with that being a ratio of the other two).  I've used @jdtournier's code snippet for the FOD field, which gives:

`-afd`
![screenie0001](https://user-images.githubusercontent.com/11952694/33472160-0efe291c-d6c5-11e7-88b2-46a60ecd438e.png)

`-peak`
![screenie0002](https://user-images.githubusercontent.com/11952694/33472163-1444a19e-d6c5-11e7-8f4c-383f83a71fc1.png)

`-disp`
![screenie0003](https://user-images.githubusercontent.com/11952694/33472170-1d18b698-d6c5-11e7-9b01-bc98ce0a830c.png)

Same ones from a "sagittal" point of view along a column where the FODs are "in plane" sagittally:

`-afd`
![screenie0004](https://user-images.githubusercontent.com/11952694/33472198-46474fe8-d6c5-11e7-8866-7e4349d3a3ab.png)

`-peak`
![screenie0005](https://user-images.githubusercontent.com/11952694/33472206-4f2a74aa-d6c5-11e7-8816-fa2f9a2bc13e.png)

`-disp`
![screenie0006](https://user-images.githubusercontent.com/11952694/33472212-5595e8b0-d6c5-11e7-80ea-e66834ed5b5e.png)

So `-afd` only varies less than 1%; the precision is inherently limited by the direction set used during that one's computation, I suppose.  This being the sharpest FODs (SH delta functions) for lmax=8, it's actually the most challenging case with respect to precision.  So for typical non-negative lmax=8 FODs, it should be even better.  But most importantly: no bias, so nothing to worry about.

`-peak` is biased in function of inclination angle (theta), but not in function of azimuth (phi).  `-peak` is only correct when the peak aligns exactly along the z-axis, and quickly starts to be biased upwards when deviating from the z-axis, up to a factor of about ~1.38.

`-disp` is the ratio of the former two, including both effects of the former two (small variance due to precision, strong bias due to `-peak` bias).  It's hence underestimated for anything deviating from the z-axis.

**Comment by jdtournier** [comment]
OK, having looked into this, I reckon there's an issue [here](https://github.com/MRtrix3/mrtrix3/blob/master/src/dwi/fmls.cpp#L266), in the [Math::SH::get_peaks()](https://github.com/MRtrix3/mrtrix3/blob/master/core/math/SH.h#L484-L529) function (most likely the [Math::SH::derivatives() call](https://github.com/MRtrix3/mrtrix3/blob/master/core/math/SH.h#L538-L622)... I'll have to go through the maths again to see if I can figure it out.

**Comment by jdtournier** [comment]
Oh, and thanks for clarifying how fixels work... :neutral_face: Definite evidence that I'm now straying into old dog territory...

**Comment by jdtournier** [comment]
OK, no it's the SH precomputer, it seems. Everything checks out if it's not used (slower, but more accurate). Not sure why it's affected here, I'll keep digging...

**Comment by jdtournier** [comment]
OK, try the fix in #1206 - I think that should fix it... Silly mistake, surprised we hadn't caught it till now. It will also impact on any SH-based tractography - although hopefully not by a huge amount. Still, not great...

**Comment by bjeurissen** [comment]
Yeah, I was worried about tracking too...

Good catch @thijsdhollander
